From 987dc095cb1f5cbf5db60c60dcd342d3700eaabd Mon Sep 17 00:00:00 2001
From: lfdevs <lfdevs@lfdevs.com>
Date: Wed, 5 Aug 2026 12:27:59 +0800
Subject: [PATCH 23/27] wayland: Relax Xwayland socket ownership for Anland

PRoot shared TMP directories can expose /tmp/.X11-unix with a UID that cannot be changed from the container. Mutter's ownership check then rejects Xwayland before it can start.

Allow the Anland backend to use such a directory while retaining the writable and sticky-bit checks. Keep ownership enforcement for other backends.

Gbp-Pq: Topic anland
Gbp-Pq: Name 0015-wayland-Relax-Xwayland-socket-ownership-for-Anland.patch

diff --git a/src/wayland/meta-xwayland.c b/src/wayland/meta-xwayland.c
index 9d7ca6a179..808f7fc17c 100644
--- a/src/wayland/meta-xwayland.c
+++ b/src/wayland/meta-xwayland.c
@@ -50,6 +50,9 @@
 
 #include "backends/meta-monitor-manager-private.h"
 #include "backends/meta-settings-private.h"
+#ifdef HAVE_ANLAND
+#include "backends/anland/meta-backend-anland.h"
+#endif
 #include "meta/main.h"
 #include "meta/meta-backend.h"
 #include "mtk/mtk-x11.h"
@@ -513,7 +516,8 @@ meta_xwayland_override_display_number (int number)
 }
 
 static gboolean
-ensure_x11_unix_perms (GError **error)
+ensure_x11_unix_perms (GError  **error,
+                       gboolean  allow_untrusted_owner)
 {
   /* Try to detect systems on which /tmp/.X11-unix is owned by neither root nor
    * ourselves because in that case the owner can take over the socket we create
@@ -545,7 +549,8 @@ ensure_x11_unix_perms (GError **error)
   /* If the directory already exists, it should belong to the same
    * user as /tmp or belong to ourselves ...
    * (if /tmp is not owned by root or ourselves we're in deep trouble) */
-  if (x11_tmp.st_uid != tmp.st_uid && x11_tmp.st_uid != getuid ())
+  if (!allow_untrusted_owner &&
+      x11_tmp.st_uid != tmp.st_uid && x11_tmp.st_uid != getuid ())
     {
       g_set_error (error, G_IO_ERROR, G_IO_ERROR_PERMISSION_DENIED,
                    "Wrong ownership for directory \"%s\", owned by %d but "
@@ -577,12 +582,13 @@ ensure_x11_unix_perms (GError **error)
 }
 
 static gboolean
-ensure_x11_unix_dir (GError **error)
+ensure_x11_unix_dir (GError  **error,
+                     gboolean  allow_untrusted_owner)
 {
   if (mkdir (X11_TMP_UNIX_DIR, 01777) != 0)
     {
       if (errno == EEXIST)
-        return ensure_x11_unix_perms (error);
+        return ensure_x11_unix_perms (error, allow_untrusted_owner);
 
       g_set_error (error, G_IO_ERROR, g_io_error_from_errno (errno),
                    "Failed to create directory \"%s\": %s",
@@ -623,12 +629,13 @@ static gboolean
 choose_xdisplay (MetaXWaylandManager     *manager,
                  MetaXWaylandConnection  *connection,
                  int                     *display,
+                 gboolean                 allow_untrusted_owner,
                  GError                 **error)
 {
   int number_of_tries = 0;
   char *lock_file = NULL;
 
-  if (!ensure_x11_unix_dir (error))
+  if (!ensure_x11_unix_dir (error, allow_untrusted_owner))
     return FALSE;
 
   do
@@ -1089,8 +1096,13 @@ meta_xwayland_init (MetaXWaylandManager    *manager,
   MetaMonitorManager *monitor_manager =
     meta_backend_get_monitor_manager (backend);
   MetaX11DisplayPolicy policy;
+  gboolean allow_untrusted_owner = FALSE;
   int display = 0;
 
+#ifdef HAVE_ANLAND
+  allow_untrusted_owner = META_IS_BACKEND_ANLAND (backend);
+#endif
+
   if (display_number_override != -1)
     display = display_number_override;
   else if (g_getenv ("RUNNING_UNDER_GDM"))
@@ -1098,11 +1110,13 @@ meta_xwayland_init (MetaXWaylandManager    *manager,
 
   if (!manager->public_connection.name)
     {
-      if (!choose_xdisplay (manager, &manager->public_connection, &display, error))
+      if (!choose_xdisplay (manager, &manager->public_connection, &display,
+                            allow_untrusted_owner, error))
         return FALSE;
 
       display++;
-      if (!choose_xdisplay (manager, &manager->private_connection, &display, error))
+      if (!choose_xdisplay (manager, &manager->private_connection, &display,
+                            allow_untrusted_owner, error))
         return FALSE;
 
       if (!prepare_auth_file (manager, error))
